packages/runtime/src/helpers/functions.ts

```diff
@@ -208,6 +227,158 @@ export const setupImageFunction = async ({
   }
 }
 
+const traceNextPages = async (pages: string[], publish: string): Promise<string[]> => {
+  const requiredFiles = await Promise.all(
+    pages.map(async (page) => {
+      const path = join(publish, 'server', 'pages', `${page}.js`)
+      const dependencies = await getDependenciesOfFile(path)
+      return [path, ...dependencies]
+    }),
+  )
+
+  return requiredFiles.flat(1)
+}
+
+const traceRequiredServerFiles = async (publish: string): Promise<string[]> => {
+  const requiredServerFilesPath = join(publish, 'required-server-files.json')
+  const {
+    files,
+    relativeAppDir,
+    config: {
+      experimental: { outputFileTracingRoot },
+    },
+  } = (await readJSON(requiredServerFilesPath)) as {
+    files: string[]
+    relativeAppDir: string
+    config: { experimental: { outputFileTracingRoot: string } }
+  }
+  const appDirRoot = join(outputFileTracingRoot, relativeAppDir)
+  const absoluteFiles = files.map((file) => join(appDirRoot, file))
+
+  absoluteFiles.push(requiredServerFilesPath)
+
+  return absoluteFiles
+}
+
+const traceNextServer = async (publish: string): Promise<string[]> => {
+  const nextServerDeps = await getDependenciesOfFile(join(publish, 'next-server.js'))
+
+  // during testing, i've seen `next-server` contain only one line.
+  // this is a sanity check to make sure we're getting all the deps.
+  if (nextServerDeps.length < 10) {
+    console.error(nextServerDeps)
+    throw new Error("next-server.js.nft.json didn't contain all dependencies.")
+  }
+
+  const filtered = nextServerDeps.filter((f) => {
+    // NFT detects a bunch of large development files that we don't need.
+    if (f.endsWith('.development.js')) return false
+
+    // not needed for API Routes!
+    if (f === 'node_modules/sass/sass.dart.js') return false
```

Comment:
But I guess in that case we don't really want to exclude it? 🤔

Reply:
good catch! in monorepos, this sometimes is `../../node_modules/sass/...`, so checking this with `.endsWith` is indeed better. Done in https://github.com/netlify/next-runtime/pull/2058/commits/b799a19580207626c5822bd2f0cefbf042c0087b.

> Is there a way that this dependency can show up in a different location in the modules tree that this check wouldn't catch, like as a transitive dependency?

This code is looking at `next-server.js.nft.json`, which contains only dependencies needed by Next.js. My understanding is that it won't contain any user code dependencies, these can be found in e.g. `pages/my-custom-page.js.nft.json`. So if we're seeing `sass/sass.dart.js` here, we know that it's because it's a direct dependency of Next.js. We don't have any evidence that it's *actually* required at runtime, though, and since it weighs 4.4mb we can exclude it here.

That said, if a user page like `pages/my-custom-page.js.nft.json` has this as it's dependency, we should definitely keep including it. The same holds true for SSR lambdas - but pure API lambdas probably don't need it.